kyo-prelude/shared/src/main/scala/kyo2/Result.scala

```diff
@@ -4,134 +4,159 @@ import Result.*
+    given [E, A](using CanEqual[A, A]): CanEqual[Result[E, A], Result[E, A]] = CanEqual.derived
+    given [E, A]: CanEqual[Result[E, A], Panic]                              = CanEqual.derived
 
-    def apply[A](expr: => A): Result[A] =
+    def apply[A](expr: => A): Result[Nothing, A] =
         try
             Success(expr)
         catch
-            case ex: Throwable if NonFatal(ex) => Failure(ex)
+            case ex: Throwable if NonFatal(ex) => Panic(ex)
 
-    def success[A](value: A): Result[A] = Success(value)
+    def success[E, A](value: A): Result[E, A] = Success(value)
 
-    def failure[A](exception: Throwable): Result[A] = Failure(exception)
+    def failure[E, A](error: E): Result[E, A] = Failure(error)
 
-    def fromEither[A](either: Either[Throwable, A]): Result[A] = either match
+    def panic[E, A](exception: Throwable): Result[E, A] = Panic(exception)
+
+    def fromEither[E, T](either: Either[E, T]): Result[E, T] = either match
         case Right(value)    => Success(value)
         case Left(exception) => Failure(exception)
 
-    opaque type Success[+A] = A | SuccessFailure[A]
+    opaque type Success[+T] = T | SuccessFailure[T]
 
     object Success:
 
-        def apply[A](value: A): Success[A] =
+        def apply[T](value: T): Success[T] =
             value match
-                case v: SuccessFailure[?]     => v.nest.asInstanceOf[Success[A]]
-                case v: Failure[A] @unchecked => SuccessFailure(v)
+                case v: SuccessFailure[?]     => v.nest.asInstanceOf[Success[T]]
+                case v: Failure[T] @unchecked => SuccessFailure(v)
                 case v                        => v
 
         // TODO avoid Option allocation
-        def unapply[A](self: Result[A]): Option[A] =
-            self.fold(_ => None)(Some(_))
+        def unapply[E, T](self: Result[E, T]): Option[T] =
+            self.fold(_ => None)(_ => None)(Some(_))
 
     end Success
 
-    case class Failure[+A](exception: Throwable)
+    sealed abstract class Cause[+E]
 
-    object Failure:
+    case class Failure[+E](error: E) extends Cause[E]
 
-        // TODO avoid Option allocation
-        def unapply[A](self: Result[A]): Option[Throwable] =
-            self.fold(Some(_))(_ => None)
+    object Failure:
+        def unapply[E, A](result: Result[E, A]): Option[E] =
+            (result: @unchecked) match
+                case result: Failure[E] => Some(result.error)
+                case _                  => None
     end Failure
 
-    extension [A](self: Result[A])
-        inline def isSuccess: Boolean =
+    case class Panic(exception: Throwable) extends Cause[Nothing]
+
+    extension [E, T](self: Result[E, T])
+
+        def isSuccess: Boolean =
             self match
-                case _: Failure[?] => false
-                case _             => true
+                case _: Cause[?] => false
+                case _           => true
 
-        inline def isFailure: Boolean = !isSuccess
```

Comment:
A few the methods don't benefit much from inlining since they don't invoke functions so I removed `inline` for them.